nixos/modules/services/libeufin/common.nix

```diff
@@ -0,0 +1,161 @@
+      dbName =
+        lib.removePrefix "postgresql:///"
+          cfg.settings."libeufin-${libeufinComponent}db-postgres".CONFIG;
+
+      bankPort = cfg.settings."${if isNexus then "nexus-httpd" else "libeufin-bank"}".PORT;
+
+      serviceDependencies =
+        # The bank and nexus shouldn't initialise the database at the same time
+        # This is only enforcable if both services are running on the same machine
+        (lib.optionals (isNexus && cfgMain.bank.enable) [ "libeufin-bank-dbinit.service" ])
+        ++ (lib.optionals cfg.createLocalDatabase [ "postgresql.service" ]);
+    in
+    lib.mkIf cfg.enable {
+      services.libeufin.settings = cfg.settings;
+
+      # TODO add system-libeufin.slice?
+      systemd.services = {
+        # Main service
+        "${serviceName}" = {
+          serviceConfig = {
+            DynamicUser = true;
+            StateDirectory = serviceName;
+            StateDirectoryMode = "0750";
+            ExecStart =
+              let
+                args = lib.cli.toGNUCommandLineShell { } {
+                  c = cfgMain.configFile;
+                  L = if cfg.debug then "debug" else null;
+                };
+              in
+              "${lib.getExe' cfg.package "libeufin-${libeufinComponent}"} serve ${args}";
+          };
+          requires = [ "${serviceName}-dbinit.service" ];
+          after = [ "${serviceName}-dbinit.service" ];
+          wantedBy = [ "multi-user.target" ];
+
+          # TODO: Could this preStart block be part of the db-init service, and then
+          #       we add it as a dependency? It seems to be part of database initialization
+          #       in a certain sense
+
+          # Some accounts might need to be registered before the
+          # service starts, like the exchange when the bank's currency
+          # conversion is enabled.
+          preStart =
+            let
+              initialAccountRegistration = lib.concatMapStringsSep "\n" (
+                account:
+                let
+                  args = lib.cli.toGNUCommandLineShell { } {
+                    c = cfgMain.configFile;
+                    inherit (account) username password name;
+                    payto_uri = "payto://x-taler-bank/bank:${toString bankPort}/${account.username}?receiver-name=${account.name}";
+                    exchange = lib.toLower account.username == "exchange";
+                  };
+                in
+                "${lib.getExe' cfg.package "libeufin-bank"} create-account ${args}"
+              ) cfg.initialAccounts;
+            in
+            lib.mkIf ((!isNexus) && (cfg.initialAccounts != [ ])) ''
+              # only register initial accounts once
+              if [ ! -e /var/lib/${serviceName}/init ]; then
+                ${initialAccountRegistration}
+
+                touch /var/lib/${serviceName}/init
+                echo "Bank initialisation complete"
+              fi
+            '';
```

Comment:
Would it make sense to move this block to `"${serviceName}-dbinit"`, so all of the database initialization is collected into one place?

Reply:
It does make sense, but in that case the service should always start, even when both databases are remote and no initialization happens.

Reply:
I'm not sure I understand your answer correctly. The `requires = [ "${serviceName}-dbinit.service" ]` and `after = [ "${serviceName}-dbinit.service" ]` should ensure that the services only can run when the database initialization has taken place and exits successfully.

It's a bit awkward to store a file on a local server stating that a remote database has been initialized, but that's a separate issue.

Reply:
> The requires = [ "${serviceName}-dbinit.service" ] and after = [ "${serviceName}-dbinit.service" ] should ensure that the services only can run when the database initialization has taken place and exits successfully.

Yes, but what I meant is that if we use remote databases for the bank and nexus, the dbinit service doesn't start since it depends on the postgresql service, right?

```nix
serviceDependencies =
  # The bank and nexus shouldn't initialise the database at the same time
  # This is only enforcable if both services are running on the same machine
  (lib.optionals (isNexus && cfgMain.bank.enable) [ "libeufin-bank-dbinit.service" ])
  ++ (lib.optionals cfg.createLocalDatabase [ "postgresql.service" ]);
```

That's why I changed it to this:

```nix
requires = lib.optionals cfg.createLocalDatabase [ "postgresql.service" ];
after = [ "network.target" ] ++ lib.optionals cfg.createLocalDatabase [ "postgresql.service" ];
```

**PS:** database initialization is now done from the same service, which is much simpler.

> It's a bit awkward to store a file on a local server stating that a remote database has been initialized, but that's a separate issue.

Yeah, but I didn't find any better solution than this.

Reply:
> [...] dbinit service doesn't start since it depends on the postgresql service, right?

As long as the dependencies of the `dbinit` service is set correctly (which they are with the `createLocalDatabase` conditionals), I think everything should be okay. The requires from the normal service unit will ensure this one is started beforehand as well.

> **PS:** database initialization is now done from the same service, which is much simpler.

Great, that was what I was asking for!